File: app/src/main/java/com/apps/agent/AgentConversationRepository.java
package com.apps.agent;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/** Separate local database: agent history is not included in account or WebDAV sync. */
public final class AgentConversationRepository {
    private static final String THREAD = "default";
    private static final int MAX_MESSAGE_CHARS = 128 * 1024;
    private static final int MAX_STORED_MESSAGES = 500;
    private final Helper helper;

    public AgentConversationRepository(Context context) {
        helper = new Helper(context.getApplicationContext());
    }

    public static final class Message {
        public final long id;
        public final String role;
        public String content;
        public String name;
        public final long createdAt;

        public Message(long id, String role, String content, String name, long createdAt) {
            this.id = id;
            this.role = role == null ? "" : role;
            this.content = content == null ? "" : content;
            this.name = name == null ? "" : name;
            this.createdAt = createdAt;
        }
    }

    public long add(String role, String content, String name) {
        ContentValues values = new ContentValues();
        values.put("thread_id", THREAD);
        values.put("role", role == null ? "" : role);
        String safeContent = content == null ? "" : content;
        if (safeContent.length() > MAX_MESSAGE_CHARS) safeContent = safeContent.substring(0, MAX_MESSAGE_CHARS) + "…";
        values.put("content", safeContent);
        values.put("name", name == null ? "" : name);
        values.put("created_at", System.currentTimeMillis());
        SQLiteDatabase db = helper.getWritableDatabase();
        long id = db.insertOrThrow("agent_messages", null, values);
        db.execSQL("DELETE FROM agent_messages WHERE thread_id=? AND id NOT IN " +
                        "(SELECT id FROM agent_messages WHERE thread_id=? ORDER BY id DESC LIMIT ?)",
                new Object[]{THREAD, THREAD, MAX_STORED_MESSAGES});
        return id;
    }

    public List<Message> recent(int limit) {
        return recentInternal(limit, false);
    }

    public List<Message> recentConversation(int limit) {
        return recentInternal(limit, true);
    }

    private List<Message> recentInternal(int limit, boolean conversationOnly) {
        int safeLimit = Math.max(1, Math.min(200, limit));
        List<Message> result = new ArrayList<>();
        Cursor cursor = helper.getReadableDatabase().query(
                "agent_messages", new String[]{"id", "role", "content", "name", "created_at"},
                conversationOnly ? "thread_id=? AND role IN ('user','assistant')" : "thread_id=?",
                new String[]{THREAD}, null, null, "id DESC", String.valueOf(safeLimit));
        try {
            while (cursor.moveToNext()) {
                String role = cursor.getString(1);
                String content = sanitizeLegacyNullPrefix(role, cursor.getString(2));
                result.add(new Message(cursor.getLong(0), role, content,
                        cursor.getString(3), cursor.getLong(4)));
            }
        } finally {
            cursor.close();
        }
        Collections.reverse(result);
        return result;
    }

    static String sanitizeLegacyNullPrefix(String role, String content) {
        String value = content == null ? "" : content;
        if (!"assistant".equals(role)) return value;
        int offset = 0;
        int count = 0;
        while (value.startsWith("null", offset)) { offset += 4; count++; }
        return count >= 2 ? value.substring(offset) : value;
    }

    public void delete(long id) {
        if (id > 0) helper.getWritableDatabase().delete("agent_messages", "id=?", new String[]{String.valueOf(id)});
    }

    public void clear() {
        helper.getWritableDatabase().delete("agent_messages", "thread_id=?", new String[]{THREAD});
    }

    private static final class Helper extends SQLiteOpenHelper {
        Helper(Context context) { super(context, "rinne_local_agent.db", null, 1); }

        @Override public void onCreate(SQLiteDatabase db) {
            db.execSQL("CREATE TABLE agent_messages (" +
                    "id INTEGER PRIMARY KEY AUTOINCREMENT," +
                    "thread_id TEXT NOT NULL," +
                    "role TEXT NOT NULL," +
                    "content TEXT NOT NULL," +
                    "name TEXT NOT NULL DEFAULT ''," +
                    "created_at INTEGER NOT NULL)");
            db.execSQL("CREATE INDEX idx_agent_messages_thread ON agent_messages(thread_id,id)");
        }

        @Override public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) { }
    }
}

File: app/src/main/java/com/apps/agent/AgentConversationRepository.kt
package com.apps.agent

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper

/** Separate local database: agent history is not included in account or WebDAV sync. */
class AgentConversationRepository(context: Context) {
    private val helper = Helper(context.applicationContext)

    class Message constructor(
        id: Long,
        role: String?,
        content: String?,
        name: String?,
        createdAt: Long
    ) {
        @JvmField val id: Long = id
        @JvmField val role: String = role ?: ""
        @JvmField var content: String = content ?: ""
        @JvmField var name: String = name ?: ""
        @JvmField val createdAt: Long = createdAt
    }

    fun add(role: String?, content: String?, name: String?): Long {
        val values = ContentValues()
        values.put("thread_id", THREAD)
        values.put("role", role ?: "")
        var safeContent = content ?: ""
        if (safeContent.length > MAX_MESSAGE_CHARS) safeContent = safeContent.substring(0, MAX_MESSAGE_CHARS) + "…"
        values.put("content", safeContent)
        values.put("name", name ?: "")
        values.put("created_at", System.currentTimeMillis())
        val db = helper.writableDatabase
        val id = db.insertOrThrow("agent_messages", null, values)
        db.execSQL(
            "DELETE FROM agent_messages WHERE thread_id=? AND id NOT IN " +
                    "(SELECT id FROM agent_messages WHERE thread_id=? ORDER BY id DESC LIMIT ?)",
            arrayOf(THREAD, THREAD, MAX_STORED_MESSAGES)
        )
        return id
    }

    fun recent(limit: Int): List<Message> = recentInternal(limit, false)

    fun recentConversation(limit: Int): List<Message> = recentInternal(limit, true)

    private fun recentInternal(limit: Int, conversationOnly: Boolean): List<Message> {
        val safeLimit = maxOf(1, minOf(200, limit))
        val result = ArrayList<Message>()
        val selection = if (conversationOnly) "thread_id=? AND role IN ('user','assistant')" else "thread_id=?"
        val cursor = helper.readableDatabase.query(
            "agent_messages", arrayOf("id", "role", "content", "name", "created_at"),
            selection, arrayOf(THREAD), null, null, "id DESC", safeLimit.toString()
        )
        cursor.use {
            while (it.moveToNext()) {
                val role = it.getString(1)
                val content = sanitizeLegacyNullPrefix(role, it.getString(2))
                result.add(Message(it.getLong(0), role, content, it.getString(3), it.getLong(4)))
            }
        }
        result.reverse()
        return result
    }

    fun delete(id: Long) {
        if (id > 0) helper.writableDatabase.delete("agent_messages", "id=?", arrayOf(id.toString()))
    }

    fun clear() {
        helper.writableDatabase.delete("agent_messages", "thread_id=?", arrayOf(THREAD))
    }

    private class Helper(context: Context) : SQLiteOpenHelper(context, "rinne_local_agent.db", null, 1) {
        override fun onCreate(db: SQLiteDatabase) {
            db.execSQL(
                "CREATE TABLE agent_messages (" +
                        "id INTEGER PRIMARY KEY AUTOINCREMENT," +
                        "thread_id TEXT NOT NULL," +
                        "role TEXT NOT NULL," +
                        "content TEXT NOT NULL," +
                        "name TEXT NOT NULL DEFAULT ''," +
                        "created_at INTEGER NOT NULL)"
            )
            db.execSQL("CREATE INDEX idx_agent_messages_thread ON agent_messages(thread_id,id)")
        }

        override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {}
    }

    companion object {
        private const val THREAD = "default"
        private const val MAX_MESSAGE_CHARS = 128 * 1024
        private const val MAX_STORED_MESSAGES = 500

        @JvmStatic
        fun sanitizeLegacyNullPrefix(role: String?, content: String?): String {
            val value = content ?: ""
            if (role != "assistant") return value
            var offset = 0
            var count = 0
            while (value.startsWith("null", offset)) { offset += 4; count++ }
            return if (count >= 2) value.substring(offset) else value
        }
    }
}
